include/boost/url/bnf/parse.hpp

```diff
@@ -77,214 +65,158 @@ parse(
-   elements [arg_0, arg_1, ..., arg_n] in @ref args.
-
-   For each type Arg_i in [Arg_0, Arg_1, ..., Arg_n] in @ref Args, a
-   corresponding function overload
-   `bool parse(char const*& it, char const* const end, error_code&, Arg_i const& t)`
-   should be defined to consume the input from [@ref it, @ref end) into `arg_i`.
-
-   This `parse` function is defined for `char`s so that it matches the BNF
-   string literals in [@ref it, @ref end).
-
-   The function returns `true` and resets @ref ec when the input is
-   parsed successfully. If the input cannot be parsed into the BNF elements,
-   the function returns `false` and sets the error number in @ref ec to the
-   appropriate error number defined in @ref boost::urls::error.
+   range of chars [`it`, `end`) into the Backus–Naur form (BNF)
+   elements in `args`. For each `arg` in `args`, a  corresponding function overload of `parse`
+   `bool parse(char const*& it, char const* const end, error_code&, Arg const& arg)`
+   should be defined to consume the input from [`it`, `end`).
 
    @par Example
    @code
-   #include <iostream>
-   #include <string>
-
-   #include <boost/url/bnf/char_set.hpp>
-   #include <boost/url/bnf/parse.hpp>
-   #include <boost/url/error.hpp>
-
-   namespace boost {
-   namespace urls {
-   namespace detail {
-   // BNF element to parse a non-empty sequence of alpha chars ([a-zA-Z]+)
-   struct alpha_string_bnf {
-     // reference to a std::string where we should store the results
-     std::string &v;
-
-     // parsing an alpha_string_bnf element
-     friend bool parse(char const *&it, char const *end, error_code &ec,
-                       alpha_string_bnf &&t) {
-       // is_alpha functor
-       bnf::alpha_chars is_alpha;
-
-       // check if empty
-       if (it == end) {
-         ec = error::incomplete;
-         return false;
-       }
-
-       // check if not alpha at all
-       if (!is_alpha(*it)) {
-         ec = error::bad_alpha;
-         return false;
-       }
-       t.v.push_back(*it);
-       ++it;
-
-       // consume more alphas
-       while (it != end && is_alpha(*it)) {
-         t.v.push_back(*it);
-         ++it;
-       }
-       ec = error::success;
-       return true;
-     }
-   };
-   } // namespace detail
-   } // namespace urls
-   } // namespace boost
-
-   int main() {
-     using namespace boost::urls;
-
-     // Input
-     std::string str = "Hello.cpp";
-     auto it = str.data();
-     auto const end = it + str.size();
-
-     // Output
-     std::string alpha_l;
-     std::string alpha_r;
-     error_code ec;
-
-     // Parse two alpha strings separated by a '.'
-     if (bnf::parse(it, end, ec, detail::alpha_string_bnf{alpha_l}, '.',
-                    detail::alpha_string_bnf{alpha_r})) {
-       std::cout << alpha_l << std::endl;
-       std::cout << alpha_r << std::endl;
-     }
-     return 0;
+   using bnf::parse;
+   if (parse(it, end, ec, bnf_element_1, bnf_element_2, bnf_element_3)) {
```

Comment:
ugh bnf_element_1... too long! This makes it hard to read and distinguish variables from types. How about `t1`, `t2`, and `t3`?

Reply:
we need to work on this

Reply:
> ugh bnf_element_1... too long! This makes it hard to read and distinguish variables from types

Maybe `bnf1`, `bnf2`, and `bnf3`?

> ugh bnf_element_1... too long! This makes it hard to read and distinguish variables from types. How about t1, t2, and t3?

From the example only, how is the user going to know [t1, t2, t3] are bnf elements?